### ISSUE TITLE
Block sharded tensor reallocation in SD is incorrect

### ISSUE THREAD

**Issue by esmalTT** [tenstorrent/tt-metal#16169]
### Summary

The following `ttnn.reallocate` in Stable Diffusion is failing with bad PCC:

https://github.com/tenstorrent/tt-metal/blob/a8a812beba9230df1196fe9ac11bb7769f5f1d90/models/demos/wormhole/stable_diffusion/tt/ttnn_functional_resnetblock2d_new_conv.py#L419

Inserting PCC checks into this above code will fail:

```python
        hidden_states = ttnn.reshape(
            hidden_states, (self.batch_size, 1, self.conv2_input_height * self.conv2_input_width, in_channels)
        )

        pre_reallocate = ttnn.to_torch(hidden_states)

        hidden_states = ttnn.reallocate(hidden_states)

        post_reallocate = ttnn.to_torch(hidden_states)

        from models.experimental.functional_unet.tests.common import verify_with_pcc
        verify_with_pcc(pre_reallocate, post_reallocate, 1.0) # FAILS WITH PCC=0.88

        hidden_states = ttnn.group_norm(
           ...
        )
```

I have isolated this failure in the following test:

```python
import pytest
import ttnn
import torch

from tests.ttnn.utils_for_testing import assert_with_pcc


def test_reallocate(device):
    input_shape = [2, 1, 4096, 320]
    core_grid = ttnn.CoreGrid(x=8, y=8)
    input_memory_config = ttnn.create_sharded_memory_config(
        input_shape, core_grid, ttnn.ShardStrategy.BLOCK, ttnn.ShardOrientation.COL_MAJOR
    )
    torch_input_tensor = torch.rand(input_shape)
    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT)

    dummy_tensor = torch.rand([1, 1, 1024, 1024])
    dummy_tensor = ttnn.from_torch(dummy_tensor, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT)
    dummy_tensor = ttnn.to_device(dummy_tensor, device, ttnn.L1_MEMORY_CONFIG)

    dummy_tensor2 = torch.rand([1, 1, 4096, 320])
    dummy_tensor2 = ttnn.from_torch(dummy_tensor2, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT)
    dummy_tensor2 = ttnn.to_device(dummy_tensor2, device, ttnn.L1_MEMORY_CONFIG)

    dummy_tensor3 = torch.rand([2, 1, 4096, 300])
    dummy_tensor3 = ttnn.from_torch(dummy_tensor3, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT)
    dummy_tensor3 = ttnn.to_device(dummy_tensor3, device, ttnn.L1_MEMORY_CONFIG)

    input_tensor = ttnn.to_device(input_tensor, device, input_memory_config)

    ttnn.deallocate(dummy_tensor3)

    output_tensor = ttnn.reallocate(input_tensor) 

    assert_with_pcc(torch_input_tensor, ttnn.to_torch(output_tensor), 1.0) # Fails with PCC=0.88

```

We can close this issue when the above test is passing.

This is blocking the following issue:
- https://github.com/tenstorrent/tt-metal/issues/16057

**Comment by esmalTT** [comment]
The issue is also present in the following test:

```python
import pytest
import ttnn
import torch

from tests.ttnn.utils_for_testing import assert_with_pcc


def test_reallocate_small(device):
    input_shape = [1, 1, 32, 40] # Any dim % 32 != 0 causes this to fail
    core_grid = ttnn.CoreGrid(x=1, y=1)
    input_memory_config = ttnn.create_sharded_memory_config(
        input_shape, core_grid, ttnn.ShardStrategy.BLOCK, ttnn.ShardOrientation.COL_MAJOR
    )
    torch_input_tensor = torch.rand(input_shape)
    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT)

    dummy_tensor = torch.rand([1, 1, 32, 32]) # Size doesn't matter
    dummy_tensor = ttnn.from_torch(dummy_tensor, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT)
    dummy_tensor = ttnn.to_device(dummy_tensor, device, ttnn.L1_MEMORY_CONFIG)

    input_tensor = ttnn.to_device(input_tensor, device, input_memory_config)
    print(input_tensor.memory_config())

    ttnn.deallocate(dummy_tensor)

    ttnn.dump_device_memory_state(device, "test_before_")
    output_tensor = ttnn.reallocate(input_tensor)
    ttnn.dump_device_memory_state(device, "test_after_")

    assert_with_pcc(torch_input_tensor, ttnn.to_torch(output_tensor), 1.0)

```

It seems like there are failures with any unaligned shard shapes.

**Comment by esmalTT** [comment]
Running the demo on top of these changes:
- https://github.com/tenstorrent/tt-metal/pull/16192

![Image](https://github.com/user-attachments/assets/6e0356f6-cb57-4d99-9b2f-973b00c928d8)

Will close once it is landed.

**Comment by esmalTT** [comment]
Fixed on main in 936cac3f9bfcd15184f7000cd732a57e8d65132e
